ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2837,264 @@ public sealed class Bot : IAsyncDisposable {
+					ArchiLogger.LogNullError(nameof(appID));
+
+					return null;
+				}
+
+				result.Add(appID);
+			}
+
+			return result;
+		}
+
+		private bool CompleteSetLootScheduled;
+
+		private async Task SendCompletedSets() {
+			lock (SendCompleteSetsSemaphore) {
+				if (CompleteSetLootScheduled) {
+					return;
+				}
+
+				CompleteSetLootScheduled = true;
+			}
+
+			await SendCompleteSetsSemaphore.WaitAsync().ConfigureAwait(false);
+
+			try {
+				lock (SendCompleteSetsSemaphore) {
+					CompleteSetLootScheduled = false;
+				}
+
+				HashSet<uint>? appIDs = await GetPossiblyCompletedBadgeAppIDs().ConfigureAwait(false);
+
+				if ((appIDs == null) || (appIDs.Count == 0)) {
+					return;
+				}
+
+				HashSet<Steam.Asset> inventory;
+
+				try {
+					inventory = await ArchiWebHandler.GetInventoryAsync(SteamID)
+						.Where(item => item.Tradable)
+						.Where(item => (item.Type == Steam.Asset.EType.TradingCard) || (item.Type == Steam.Asset.EType.FoilTradingCard))
+						.Where(item => appIDs.Contains(item.RealAppID))
+						.ToHashSetAsync()
+						.ConfigureAwait(false);
+				} catch (HttpRequestException e) {
+					ArchiLogger.LogGenericWarningException(e);
+
+					return;
+				} catch (Exception e) {
+					ArchiLogger.LogGenericException(e);
+
+					return;
+				}
+
+				if (inventory.Count == 0) {
+					ArchiLogger.LogGenericWarning(string.Format(Strings.ErrorIsEmpty), nameof(inventory));
+
+					return;
+				}
+
+				Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), List<uint>> inventorySets = Trading.GetInventorySets(inventory);
+				List<Task<((uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity) Key, (uint Sets, byte CardsPerBadge) Value)>> loadCardCountsTasks = inventorySets.Select(async kv => (kv.Key, (Sets: kv.Value.FirstOrDefault(), CardsPerBadge: await ArchiWebHandler.GetCardCountForGame(kv.Key.RealAppID).ConfigureAwait(false)))).ToList();
+				Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), (uint Sets, byte CardsPerSet)> itemsToTakePerInventorySet = (await Utilities.InParallel(loadCardCountsTasks).ConfigureAwait(false)).ToDictionary(entry => entry.Key, entry => entry.Value);
+				HashSet<Steam.Asset> result = GetItemsForFullBadge(inventory, itemsToTakePerInventorySet);
+
+				if (result.Count > 0) {
+					await Actions.SendInventory(result).ConfigureAwait(false);
+				}
+			} finally {
+				SendCompleteSetsSemaphore.Release();
+			}
+		}
+
+		internal static HashSet<Steam.Asset> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> inventory, Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), (uint SetsToExtract, byte CardsPerSet)> amountsToExtract) {
+			if ((inventory == null) || (amountsToExtract == null)) {
+				throw new ArgumentNullException($"{nameof(inventory)} || {nameof(amountsToExtract)}");
+			}
+
+			if ((amountsToExtract.Count == 0) || (inventory.Count == 0)) {
+				return new HashSet<Steam.Asset>(0);
```

Comment:
Move this condition above and throw exception as you do, if the caller is brain-damaged to provide us with nonsense stuff, he's going to get yelled at.